app/views/sub.py

```diff
@@ -205,6 +205,12 @@ def view_sub_new(sub, page):
 
     posts = misc.getPostList(misc.postListQueryBase(noAllFilter=True).where(Sub.sid == sub['sid']),
                              'new', page).dicts()
+                             
+    for post in posts:
+        if current_user.is_mod(sub['sid'], 1) or current_user.is_admin():
+            post['open_reports'] = SubPostReport.select().where((SubPostReport.pid == post['pid']) & SubPostReport.open == True).dicts()
+        else:
+            post['open_reports'] = False
```

Comment:
For performance reasons, this should only be executed if the user is a mod (or admin), and it would be even better if it was a join in the posts query or just a single query for all the posts listed (because queries inside a loop = bad)

Reply:
I agree this would be better... but I can't figure out how to make that query :disappointed: 

here's as far as I've gotten, editing postListQueryBase in misc.py:

```
def postListQueryBase(*extra, nofilter=False, noAllFilter=False, noDetail=False, adminDetail=False, isSubMod=False):
...
if isSubMod:
     posts = posts.switch(SubPost).select(SubPostReport.id.alias('open_reports')).join(SubPostReport, JOIN.LEFT_OUTER).where(SubPostReport.pid == SubPost.pid).switch(SubPost)
```
I'm getting `KeyError: 'pid'` which makes no sense to me... Here's the traceback:
``
/home/mk/Documents/MK_Laptop_backup/mk/Projects/throat/app/misc.py:89: UserWarning: autoreload limitation: master (inherited), imported and preprocessed templates. It is recommended to use application server that supports file reload instead.
  engine = autoreload(_engine)
 * Restarting with stat
/home/mk/Documents/MK_Laptop_backup/mk/Projects/throat/app/misc.py:89: UserWarning: autoreload limitation: master (inherited), imported and preprocessed templates. It is recommended to use application server that supports file reload instead.
  engine = autoreload(_engine)
 * Debugger is active!
 * Debugger PIN: 271-763-028
Traceback (most recent call last):
  File "/home/mk/.virtualenvs/throat/lib/python3.6/site-packages/wheezy/template/engine.py", line 48, in render
    return self.renders[name](ctx, local_defs, super_defs)
KeyError: 'shared/layout.html'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mk/.virtualenvs/throat/lib/python3.6/site-packages/flask/app.py", line 2450, in wsgi_app
    response = self.handle_exception(e)
  File "/home/mk/.virtualenvs/throat/lib/python3.6/site-packages/flask/app.py", line 1867, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/mk/.virtualenvs/throat/lib/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/mk/.virtualenvs/throat/lib/python3.6/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/mk/.virtualenvs/throat/lib/python3.6/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/mk/.virtualenvs/throat/lib/python3.6/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/mk/.virtualenvs/throat/lib/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/mk/.virtualenvs/throat/lib/python3.6/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/mk/.virtualenvs/throat/lib/python3.6/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/mk/Documents/MK_Laptop_backup/mk/Projects/throat/app/views/sub.py", line 307, in view_sub_hot
    'subMods': misc.getSubMods(sub['sid'])})
  File "/home/mk/.virtualenvs/throat/lib/python3.6/site-packages/wheezy/template/engine.py", line 134, in render
    return self.render_template(ctx, {}, {})
  File "sub.html", line 56, in render
    
  File "/home/mk/.virtualenvs/throat/lib/python3.6/site-packages/wheezy/template/engine.py", line 51, in render
    return self.renders[name](ctx, local_defs, super_defs)
  File "shared/layout.html", line 150, in render
    
  File "shared/layout.html", line 141, in content
    
  File "sub.html", line 36, in main
    
KeyError: 'pid'
```

I can push my branch up as-is if that's helpful. The way these joins work is a bit of a black box to me...

Reply:
As far as I know you can't have more than one `.select()` call, so the second one is overwriting the first. We have an ugly solution for conditional fields in `get_comment_tree`. You can find it with the comment `silly hack`.

I'm still trying to look up better solutions for that though

Reply:
Oh I see, thanks. I'll try something along those lines

Reply:
Okay update. So this way is "working" in the sense that there are no errors.... but I think it's actually filtering my the posts with open reports rather than returning the ids of the open reports as a list, which is what I was aiming for...

```
def postListQueryBase(*extra, nofilter=False, noAllFilter=False, noDetail=False, adminDetail=False, isSubMod=False):
    if current_user.is_authenticated and not noDetail:
        posts = SubPost.select(SubPost.nsfw, SubPost.content, SubPost.pid, SubPost.title, SubPost.posted,
                               SubPost.deleted, SubPost.score, SubPost.ptype,
                               SubPost.thumbnail, SubPost.link, User.name.alias('user'), Sub.name.alias('sub'),
                               SubPost.flair, SubPost.edited, Sub.sid,
                               SubPost.comments, SubPostVote.positive, User.uid, User.status.alias('userstatus'),
                               *extra, *([SubPostReport.id.alias('open_reports')] if isSubMod else [Value(None).alias('open_reports')]))
        posts = posts.join(SubPostVote, JOIN.LEFT_OUTER,
                           on=((SubPostVote.pid == SubPost.pid) & (SubPostVote.uid == current_user.uid))).switch(
            SubPost).join(SubPostReport, JOIN.LEFT_OUTER).where((SubPostReport.pid == SubPost.pid) & SubPostReport.open == True).switch(SubPost)
```

Reply:
That's because the `.where()` is doing that filtering. If you intended to make that extra where the condition for the join, it should be `.join(SubPostReport, JOIN.LEFT_OUTER, on=(condition))`

Also you shouldn't do that join unless isSubMod is true

Reply:
Okay this is progress! Now it is displaying any post with a report on it once for each report lol

![Screenshot from 2020-07-25 14 45 05](https://user-images.githubusercontent.com/17955536/88465067-7e1ca480-ce85-11ea-8726-5944fc1b6866.png)

```
def postListQueryBase(*extra, nofilter=False, noAllFilter=False, noDetail=False, adminDetail=False, isSubMod=False):
    if current_user.is_authenticated and not noDetail:
        posts = SubPost.select(SubPost.nsfw, SubPost.content, SubPost.pid, SubPost.title, SubPost.posted,
                               SubPost.deleted, SubPost.score, SubPost.ptype,
                               SubPost.thumbnail, SubPost.link, User.name.alias('user'), Sub.name.alias('sub'),
                               SubPost.flair, SubPost.edited, Sub.sid,
                               SubPost.comments, SubPostVote.positive, User.uid, User.status.alias('userstatus'),
                               *extra, *([SubPostReport.id.alias('open_reports')] if isSubMod else [Value(None).alias('open_reports')]))
        posts = posts.join(SubPostVote, JOIN.LEFT_OUTER,
                           on=((SubPostVote.pid == SubPost.pid) & (SubPostVote.uid == current_user.uid))).switch(
            SubPost)
        if isSubMod:
            posts = posts.join(SubPostReport, JOIN.LEFT_OUTER, on=((SubPostReport.pid == SubPost.pid) & (SubPostReport.open == True))).switch(SubPost)
```

Reply:
yeah you should use a `.group_by(SubPost.pid)` or the join will multiply everything

Reply:
Okay thanks, that helped solve that! Now the problem is that 'open_reports' is returning as an integer rather than a list of report ids (looks like it's just taking the id of the latest open report).

Reply:
Use the SQL count function?

Reply:
But I do need at least ONE of the IDs to link to the report so mod can take action, otherwise it's not helpful...

Ferret has suggested I do it like the comment history by getting the open_reports for ALL the posts in the query first, and then distributing them to their associated post. (as seen here: https://github.com/Phuks-co/throat/blob/master/app/misc.py#L1629)

Reply:
Yeah that can be a better approach

Reply:
Alright, thanks for your help. This seems like a tomorrow task haha